cli/resolver.rs

```diff
@@ -65,11 +72,39 @@ impl Resolver for CliResolver {
     referrer: &ModuleSpecifier,
   ) -> Result<ModuleSpecifier, AnyError> {
     if let Some(import_map) = &self.maybe_import_map {
-      import_map
+      return import_map
         .resolve(specifier, referrer)
-        .map_err(|err| err.into())
-    } else {
-      resolve_import(specifier, referrer.as_str()).map_err(|err| err.into())
+        .map_err(|err| err.into());
     }
+
+    if let Some(pjson) = self.maybe_package_json.as_ref() {
+      let maybe_specifier_and_version = if let Some(deps) = &pjson.dependencies
+      {
+        if deps.contains_key(specifier) {
+          let version = deps.get(specifier).unwrap();
+          Some((specifier, version))
+        } else {
+          None
+        }
+      } else if let Some(dev_deps) = &pjson.dev_dependencies {
+        if dev_deps.contains_key(specifier) {
+          let version = dev_deps.get(specifier).unwrap();
+          Some((specifier, version))
+        } else {
+          None
+        }
+      } else {
+        None
+      };
+
+      if let Some((specifier, version)) = maybe_specifier_and_version {
+        return Ok(
+          ModuleSpecifier::parse(&format!("npm:{specifier}@{version}"))
+            .unwrap(),
+        );
```

Comment:
For resolving bare specifiers I think we need to do the following:

1. Parse all dependencies and dev dependencies in package.json via `NpmPackageReq::from_dependency_entry(key, value)` to get a list of npm package reqs.
2. Add these package requirements into the npm resolver.
3. Get the resolved `NpmPackageReq` exact version.
4. Create a map of `key` -> to the exact version (ex. `npm:{specifier}@{version}`)... I think we'll need to inject these exact specifiers into the npm resolver as well (maybe as a first pass)
5. Use that mapping here.

Reply:
I was just thinking about this more and maybe we should be stricter with what's allowed in the current project's package.json and only allow version constraints that are supported in npm specifiers. It would simplify our implementation and not complicate the lockfile even more. I think maybe it's not a big deal to be strict because it's a package.json that the user has control over so they could easily edit their package.json and update it to work with Deno in case they do something that's not supported.